File: 234. Palindrome Linked List/234.PalindromeLinkedList.java
class Solution {
    public boolean isPalindrome(ListNode head) {
        // Create a list to store the node values for easy comparison
        List<Integer> list = new ArrayList<>();

        // Traverse the linked list and add each node's value to the list
        while(head != null) {
            list.add(head.val);
            head = head.next; // Move to the next node
        }

        // Initialize two pointers: one at the start (left) and one at the end (right)
        int left = 0;
        int right = list.size() - 1;

        // Compare values from both ends of the list moving towards the center
        // If at any point the values don't match, it's not a palindrome
        while(left < right && list.get(left) == list.get(right)) {
            left++; // Move the left pointer towards the center
            right--; // Move the right pointer towards the center
        }

        // If the pointers have crossed, it means all compared values were the same, so it's a palindrome
        return left >= right;
    }
}

File: 234. Palindrome Linked List/234.PalindromeLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
var isPalindrome = function (head) {
  // Create an array to store node values for comparison
  const list = [];

  // Traverse the linked list and push each node's value to the array
  while (head !== null) {
    list.push(head.val);
    head = head.next; // Move to the next node
  }

  // Initialize two pointers: one at the start (left) and one at the end (right)
  let left = 0;
  let right = list.length - 1;

  // Compare values from both ends of the list moving towards the center
  while (left < right && list[left] === list[right]) {
    left++; // Move the left pointer towards the center
    right--; // Move the right pointer towards the center
  }

  // If the pointers have crossed, all compared values were the same, so it's a palindrome
  return left >= right;
};
